timApp/modules/cs/js/editor/math-editor/formula-editor.component.ts

```diff
@@ -0,0 +1,1072 @@
+        }
+
+        if (this.formulaType != FormulaType.Align) {
+            this.formulaType =
+                this.fields.length > 1 ? FormulaType.Multi : FormulaType.Inline;
+        }
+        this.isDisabled = this.fields.length > 1;
+        this.useExistingParenthesis = false;
+        this.updateFormulaToEditor();
+    }
+
+    /**
+     * Sets LaTeX content of a field,
+     * updates changes to preview and
+     * sets edited field as active
+     * @param res edit content
+     */
+    handleEdited(res: Edit) {
+        if (res.id < 0 || res.id >= this.fields.length) {
+            return;
+        }
+        this.fields[res.id].latex = res.latex;
+        this.updateFormulaToEditor();
+        this.activeFieldsIndex = res.id;
+    }
+
+    /**
+     * Sets active field
+     * @param res edit content
+     */
+    handleFocus(res: Edit) {
+        this.activeFieldsIndex = res.id;
+    }
+
+    /**
+     * Sets active field as one after currently active one if one exists.
+     */
+    handleArrowDown() {
+        if (this.activeFieldsIndex + 1 < this.fields.length) {
+            this.activeFieldsIndex++;
+        }
+    }
+
+    /**
+     * Sets active field as one before currently active one if one exists.
+     */
+    handleArrowUp() {
+        if (this.activeFieldsIndex > 0) {
+            this.activeFieldsIndex--;
+        }
+    }
+
+    /**
+     * Find the line where cursor is in editor.
+     */
+    getCurrentLine() {
+        const text = this.editor.content;
+        let startI = this.cursorLocation;
+        let endI = this.cursorLocation;
+        if (text[startI] === "\n") {
+            startI--;
+        }
+        while (startI >= 0 && text[startI] !== "\n") {
+            startI--;
+        }
+        while (endI < text.length && text[endI] !== "\n") {
+            endI++;
+        }
+        return text.slice(startI + 1, endI);
+    }
+
+    /**
+     * Determine formula type based on current cursor location
+     */
+    getInitialFormulaType() {
+        const currentLine = this.getCurrentLine();
+        const isEmptyLine = currentLine.trim().length === 0;
+        // should be multiline if no real text in line
+        return isEmptyLine ? FormulaType.Multi : FormulaType.Inline;
+    }
+
+    /**
+     * Splits string into two parts if possible at cursor location.
+     */
+    parseOldContent() {
+        if (!this.editor.cursorIndexPosition) {
+            this.oldContent = {
+                before: this.editor.content,
+                editing: "",
+                after: "",
+            };
+            return;
+        } else {
+            const cursorI = this.editor.cursorIndexPosition();
+            this.oldContent = {
+                before: this.editor.content.slice(0, cursorI),
+                editing: "",
+                after: this.editor.content.slice(cursorI),
+            };
+        }
+    }
+
+    /**
+     * Method to find $ and $$ syntax parenthesis from a string.
```

Comment:
Mielellään kuvaus mallia "Finds ..." kuin että "Method that does X". Siten dokumentaatio on hieman yhtenäisempi.